src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3645,74 +3693,114 @@ protected override void VisitBinaryOperatorChildren(ArrayBuilder<BoundBinaryOper
                 binary = stack.Pop();
             }
 
-            // learn from true/false constant
-            bool learnFromBooleanConstantTest()
+            static bool isEqualsOrNotEquals(BoundBinaryOperator binary)
+                => binary.OperatorKind.Operator() is BinaryOperatorKind.Equal or BinaryOperatorKind.NotEqual;
+
+            static bool isEquals(BoundBinaryOperator binary)
+                => binary.OperatorKind.Operator() == BinaryOperatorKind.Equal;
+
+            static bool isKnownNullOrNotNull(BoundExpression expr, TypeWithState resultType)
+            {
+                return resultType.State.IsNotNull()
+                    || expr.ConstantValue is object;
+            }
+
+            LocalState getSingleState(BoundExpression otherOperand, PossiblyConditionalState conditionalStateWhenNotNull)
             {
-                if (!IsConditionalState)
+                LocalState stateWhenNotNull;
+                if (!conditionalStateWhenNotNull.IsConditionalState)
                 {
-                    return false;
+                    stateWhenNotNull = conditionalStateWhenNotNull.State;
                 }
-
-                if (!leftConversion.IsIdentity)
+                else if (isEquals(binary) && otherOperand.ConstantValue is { IsBoolean: true, BooleanValue: var boolValue })
                 {
-                    return false;
+                    // can preserve conditional state from `.TryGetValue` in `dict?.TryGetValue(key, out value) == true`,
+                    // but not in `dict?.TryGetValue(key, out value) != false`
+                    stateWhenNotNull = boolValue ? conditionalStateWhenNotNull.StateWhenTrue : conditionalStateWhenNotNull.StateWhenFalse;
```

Comment:
But what about `is true or false` (just to throw more curveballs at you :P).

Reply:
Coming in #53311 :)